### PR TITLE
UI editor for `shopping-list` card

### PR DESCRIPTION
**Pull request by iantrich**
shopping-editor

### CODE REVIEW

File: src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    return html`
+      ${configElementStyle}
+      <div class="card-config">
+        <paper-input
+          label="Title"
+          value="${this._title}"
+          .configValue="${"title"}"
+          @value-changed="${this._valueChanged}"
+        ></paper-input>
+        </div>
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts]:
```suggestion
```

File: src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    const target = ev.target! as EditorTarget;
+
+    if (this[`_${target.configValue}`] === target.value) {
+      return;
+    }
+    if (target.configValue) {
+      this._config = {
+        ...this._config,
+        [target.configValue!]:
+          target.checked !== undefined ? target.checked : target.value,
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts]: This is only used for checkboxes, not needed here.

File: src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
+import { html, LitElement, PropertyDeclarations } from "@polymer/lit-element";
+import { TemplateResult } from "lit-html";
+import { struct } from "../../common/structs/struct";
+import "@polymer/paper-dropdown-menu/paper-dropdown-menu";
+import "@polymer/paper-item/paper-item";
+import "@polymer/paper-listbox/paper-listbox";
+import "@polymer/paper-toggle-button/paper-toggle-button";
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts]: You only use paper-input, so that should be imported, the rest can go.

**Review by bramkragten — COMMENTED** [review]
Looks good! minor comments 👍

File: src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts (hunk 3 above)
Comment (iantrich) [reply src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts]: yeah, cleaning up now :)

File: src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+export class HuiShoppingListEditor extends hassLocalizeLitMixin(LitElement)
+  implements LovelaceCardEditor {
+  public hass?: HomeAssistant;
+  private _config?: Config;
+
+  public setConfig(config: Config): void {
+    config = cardConfigStruct(config);
+
+    this._config = { type: "shopping-list", ...config };
```
Comment (balloob) [inline src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts]: Isn't it weird that we have to add type ourselves as a default here? If there was no type set, it wouldn't end up in this component?

File: src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          label="Title"
+          value="${this._title}"
+          .configValue="${"title"}"
+          @value-changed="${this._valueChanged}"
+        ></paper-input>
+      </div>
+    `;
+  }
+
+  private _valueChanged(ev: EntitiesEditorEvent): void {
```
Comment (balloob) [inline src/panels/lovelace/editor/config-elements/hui-shopping-list-editor.ts]: How many of these have we written and do we already feel like it's time to generalize?

```ts
export const valueChanged = <T>({ hass: HomeAssistant, _config: T, ev: Event }) => {

} 
```

Also, this is not an EntitiesEditorEvent

**Pull request merged**